File: LeetCode/75.Sort_Colors/Mario/75.Sort_Colors.py
#!/usr/bin/env python3


class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        # three way quick sort
        # three pointer (0, 2, current)

        n = len(nums)
        # j: 0, k: 2
        i, j, k = 0, 0, n - 1
        while i <= k:
            cur = nums[i]
            if cur == 0:
                nums[i], nums[j] = nums[j], nums[i]
                j += 1
                i += 1
            elif cur == 2:
                nums[i], nums[k] = nums[k], nums[i]
                k -= 1
            else:
                i += 1

File: LeetCode/75.Sort_Colors/75.Sort_Colors.java
class Solution {
    public void sortColors(int[] nums) {
        int p0 = 0, p2 = nums.length - 1;
        int i = 0;
        while (i <= p2) {
            if (nums[i] == 0) {
                swap(nums, i, p0);
                p0 += 1;
                i += 1;
            } else if (nums[i] == 2) {
                swap(nums, i, p2);
                p2 -= 1;
            } else {
                i += 1;
            }
        }
    }

    private void swap(int[] nums, int a, int b) {
        int tmp = nums[a];
        nums[a] = nums[b];
        nums[b] = tmp;
    }
}